### PR TITLE
Add automatic translation of items to ItemList

### PR DESCRIPTION
**Pull request by DennisManaa**
## What this PR does
While using the ItemList-Node, I noticed that items were not automatically translated when switching between locales. To rectify this issue, I am submitting this pull request (PR). The prior implementation lacked this functionality entirely, so this PR introduces automatic item translation to the ItemList-Node. (Notably, it's worth mentioning that many other nodes, like MenuButton or OptionButton, already include automatic translation. Initially, I presumed the issue addressed in this PR might be a bug.)

## How it was before
[Bildschirmaufzeichnung vom 2023-10-18, 21-18-34.webm](https://github.com/godotengine/godot/assets/26871131/d52d8215-1df6-499a-ae32-eae53dc84192)

## How it is now
[Bildschirmaufzeichnung vom 2023-10-18, 21-23-23.webm](https://github.com/godotengine/godot/assets/26871131/840627b7-fa4a-40fb-a0d2-e9bdaf470a3f)

## Minimal Test Project
Here you can download the project I used in the videos above.
[test-project.zip](https://github.com/godotengine/godot/files/13033361/test-project.zip)

### CODE REVIEW

**Comment by DennisManaa** [comment]
To provide further clarity regarding the nature of this PR, I'd like to argue that it isn't precisely a "feature proposal.". However, I am open to any classification that helps the process. :)

As demonstrated in the accompanying video, numerous other nodes already exhibit the same behavior that I have introduced in this PR for the ItemList (Even the button that is used for switching the locale, uses translations to switch the text you can see in the video). The absence of automatic translations in the ItemList-Node is more of an exception. So for an end user, the absence of this feature in the ItemList feels more like a bug, because virtually all/most other GUI-Nodes support it currently.

Technically, I absolutely agree that it qualifies as an enhancement, and technically, it isn't a bug fix since automatic translation was never implemented for the ItemList. However, I'm just uncertain if it should be categorized as a "feature proposal" when all or at least the majority of other GUI-Nodes already exhibit the exact behavior that I have implemented here. 

[Bildschirmaufzeichnung vom 2023-10-18, 23-23-38.webm](https://github.com/godotengine/godot/assets/26871131/bb5ee709-db06-4e21-9bc8-d2847304fd4f)

**Comment by KoBeWi** [comment]
While the PR uses already-existing functionality, it changes the node's behavior by adding a new feature. It doesn't really matter, we had similar PRs for other nodes (like PopupMenu), so it probably doesn't need a proposal.

That said, you should check ItemList usage in the editor and call `set_auto_translate(false)` where necessary, as this change will likely cause words getting translated randomly when they shouldn't.

**Comment by DennisManaa** [comment]
Thank you for the clarification; I get the point. :)

Regarding the use of ItemList in the editor, you're absolutely right—I hadn't considered that aspect. I will investigate it as soon as possible, hopefully by tomorrow or the day after. Thanks for mentioning this :+1:

File: scene/gui/item_list.cpp
Comment (bruvzg) [inline scene/gui/item_list.cpp:1007]: Probably better to split these cases to avoid unnecessary translation lookups on unrelated theme changes.

```c++
		case NOTIFICATION_TRANSLATION_CHANGED: {
 			for (int i = 0; i < items.size(); i++) {
				items.write[i].xl_text = atr(items[i].text);
				_shape_text(i);
			}
			shape_changed = true;
			queue_redraw();
		} break;
		case NOTIFICATION_THEME_CHANGED: {
 			for (int i = 0; i < items.size(); i++) {
				_shape_text(i);
			}
			shape_changed = true;
			queue_redraw();
		} break;
````

Comment (DennisManaa) [reply scene/gui/item_list.cpp]: Thanks for the feedback! I added the change you asked for :)

**Comment by DennisManaa** [comment]
@KoBeWi, I've examined the usage of ItemList in the editor, and there are a total of 28 occurrences. The majority of these ItemLists are employed for elements such as file names, method names, or contributor names in the "About Godot" window. These are all elements that should remain unaltered in translation.

There are only a few instances where the text within the editor truly requires translation. For example, in the Export Window, the text within parentheses is translated. However, it's worth noting that this functionality worked properly in the past (and as far as I can tell when autotranslation is disabled) due to an alternate implementation (e. g. in the screenshots, where the text "Runnable" inside the parentheses is translated).

For now, I've disabled autotranslation for all instances of ItemList in the editor. However, I'm unsure whether we should replace the individual attempts at automatic translation in this pull request. Probably it would be a good idea to do this in individual pull requests, where it can be tested in an isolated manner, as such changes might carry the risk of introducing new bugs. What do you think about this matter?

grafik
grafik

**Review by QbieShay — APPROVED** [review]
Approving on behalf of shaders (changes are trivial), but still needs approval from editor maintainers.

File: scene/gui/item_list.cpp
Comment (DennisManaa) [reply scene/gui/item_list.cpp:1020]: Thanks for the feedback, I added the space :)

**Review by AThousandShips — APPROVED** [review]
Havent tested but looks straightforward

**Review by KoBeWi — APPROVED** [review]
The implementation looks ok.

That said, I wonder how much this feature is useful. Unlike other controls with list auto-translation, like MenuButton or OptionButton, ItemList is intended for dynamic content that most often shouldn't be translated. It's similar to Tree in that regard, which can't even be edited in the inspector. Maybe the `auto_translate` should be disabled by default for ItemList? It's doable by calling `set_auto_translate(false)` in the constructor. (this would make your other changes unnecessary 🙃) Up to discussion.

**Comment by DennisManaa** [comment]
Thanks for Reviewing :)

The question regarding the usefulness of this feature is indeed a good one. I'll try to explain my specific use case for this feature and why I think that it is valuable, but of course it's up to discussion if it is valuable for all users of the engine. 

Currently, I utilize ItemLists on a settings screen within my game in a manner similar to OptionButtons. Nonetheless, there exists a fundamental distinction: ItemLists consistently display all available options to the user, thereby enhancing usability, particularly in cases where the options are limited in number. (my opinion/design decision)

To illustrate, the settings screen closely resembles the one shown in the following video. There are only 2 ItemLists and the MSAA-ItemList is a very bad example, but still it explains the usecase:

[Bildschirmaufzeichnung vom 2023-10-28, 01-55-05.webm](https://github.com/godotengine/godot/assets/26871131/dac824c4-746b-4ae8-8763-d1a1ffcf1e74)

In this specific scenario, the text that requires translation remains static and never changes. Therefore, it seems logical to include it in the translation file. However, to achieve this, I currently rely on the following script:

```
extends ItemList

var _initial_option_texts: Array[String] = [];

func _ready() -> void:
    var item_num: int = 0;
    while(item_num < item_count):
        _initial_option_texts.append(get_item_text(item_num));
        item_num += 1
    
    _translate();
    SettingsManager.language_changed.connect(_translate);

func _translate() -> void:
    var item_num: int = 0;
    while(item_num < item_count):
        set_item_text(item_num, tr(_initial_option_texts[item_num]));
        item_num += 1;
```

In my personal view, this approach is suboptimal and counterintuitive, as most/all other elements support automatic translation. Initially, I even suspected a bug, when working with ItemLists, because it worked with all of the other Nodes I used.

Furthermore, I am open to discussing the possibility of disabling the `auto_translate` setting for ItemLists by default. Both options are fine to me, but I believe it's important to consider how other elements handle this to maintain consistency. Introducing an inconsistency, for instance, if all other elements enable auto translation by default, should be weighed carefully, because it might confuse users of the engine as the absence of auto-translation for ItemLists once confused me. :upside_down_face: 

So let's just change this PR in any way that fits best :)

**Comment by KoBeWi** [comment]
> Furthermore, I am open to discussing the possibility of disabling the auto_translate setting for ItemLists by default. Both options are fine to me, but I believe it's important to consider how other elements handle this to maintain consistency. Introducing an inconsistency, for instance, if all other elements enable auto translation by default, should be weighed carefully, because it might confuse users of the engine as the absence of auto-translation for ItemLists once confused me. 🙃

We do have some nodes that change the default values. Such overrides are listed in the documentation. We could add a more prominent note in the description of the class.

**Comment by DennisManaa** [comment]
> We do have some nodes that change the default values. Such overrides are listed in the documentation. We could add a more prominent note in the description of the class.

In this case, I'd be happy to go ahead with your suggestion to disable `auto_translate by default. Do you think we need another opinion on this, or should I just start implementing it? :)

**Comment by KoBeWi** [comment]
There is no rush, we could wait.
Or you can make this change in a separate commit for now.

**Review by YeldhamDev — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by CsloudX** [comment]
`Tree` & `GraphNode` need fix also. ^V^ https://github.com/godotengine/godot/issues/80549

**Pull request merged**
